### ISSUE TITLE
site-url not respected in links

### ISSUE THREAD

**Issue by mimoo** [rust-lang/mdBook#1764]
Since I set the site-url (or base url in other frameworks):

```toml
[output.html]
site-url = "/proof-systems/"
```

I would expect all markdown links to be rewritten to respect that in my book. They aren't :o

**Comment by ehuss** [comment]
All links in your book should be relative, so it shouldn't matter where it is located.  The `site-url` is only needed for generating the 404 page which has to be able to work when loaded from any location.  

If you are using absolute links, like `/foo.md`, then I recommend changing them to be relative (such as `foo.md` or `../foo.md`).  I'm not sure mdbook should rewrite absolute links, since they may legitimately be trying to link to a location outside of the book.

**Comment by mimoo** [comment]
in my case, I am moving the generated rust documentation into the `/rustdoc` folder in the book. So that on github pages, I can access both [the book](https://o1-labs.github.io/proof-systems/) and [the rustdoc](https://o1-labs.github.io/proof-systems/rustdoc/). 

> I'm not sure mdbook should rewrite absolute links, since they may legitimately be trying to link to a location outside of the book.

I'm not sure I understand, all frameworks that expose a base-url do this: they rewrite all absolute links using the base url. What is a use case where you wouldn't want that?

**Comment by schungx** [comment]
> I'm not sure I understand, all frameworks that expose a base-url do this: they rewrite all absolute links using the base url. What is a use case where you wouldn't want that?

I'd second this.

Writing a large book with `mdbook` means that I'm constantly shifting chapters from one place to another, sometimes across folders or up/down levels.  It would be a huge mess to manually fix all relative url links whenever I move a file.

Always using absolute url links resolve this issue nicely, resulting in a huge savings of time.  I believe this feature is sorely needed.

Right now, I use a templating engine `mdbook-tera` and put `{{rooturl}}/` in front of all my url links, but it is not as good as built-in support.

**Comment by ehuss** [comment]
> What is a use case where you wouldn't want that?

This probably isn't used in practice very often, but I can imagine a book that wants to link to other pages on the same website (that aren't part of a book).  We don't use it for rust-lang, so I would be fine changing it, but it could possibly be a breaking change for other users, so it would need to wait for the next incompatible release.

**Comment by mimoo** [comment]
> This probably isn't used in practice very often, but I can imagine a book that wants to link to other pages on the same website (that aren't part of a book)

yeah I'm not personally convinced, from a site-url standpoint everything else is "external" and should be linked with a fully-qualified URL

**Comment by schungx** [comment]
> This probably isn't used in practice very often

Actually it is used very often for me.

It is not _absolutely needed_ as I can always use `../../../some/path/doc.md` instead of the much better `/some/path/doc.md`, but that means that I can never move files around as well as up/down levels without doing a global scan for broken links!

**Comment by joaofreires** [comment]
Hey you all! Not sure if what I did solves your issue. But could you check #1802 and test it if you have some time?
Thank you!

**Comment by schungx** [comment]
What should I do to test an open unmerged PR?  Never done this before...

**Comment by joaofreires** [comment]
> or is it in the add_absolute_links_support branch?

Yes, you can build the `add_absolute_links_support ` branch and test from there.

Here's the repo/branch:  https://github.com/joaofreires/mdBook/tree/add_absolute_links_support 

Note: as described in the PR, you need to set the feature as `true`. So, inside the `book.toml` you will need to have something like:
```toml
[output.html]
site-url = "/some/path/"
use-site-url-as-root = true
```

**Comment by schungx** [comment]
Seems to work just fine!

I particularly like the fact that it _doesn't_ add the site url if going through `mdbook serve`, that makes things _Just Work_ everywhere.

**Comment by joaofreires** [comment]
@schungx , mdBook already set the `site-url` to `/` in the "serve mode". 😃 

https://github.com/rust-lang/mdBook/blob/7c37dd5e8570ffc93701c1d6b6ffdd97b430467e/src/cmd/serve.rs#L79

**Comment by schungx** [comment]
> @schungx , mdBook already set the `site-url` to `/` in the "serve mode". 😃

With this change, I can actually remove `mdbook-tera` because I only used the templating engine in order to put in the correct url prefix.  It makes the entire project much simpler to build and deploy!

**Comment by schungx** [comment]
Now if I would ask for one more thing: make the `#include` command take paths that are based off `src`.

Right now, it is either relative path based off the current file, or an absolute path based off the file system root (which is extremely unusual).  It is much more common to have include files offset from `src`.

**Comment by joaofreires** [comment]
@schungx, the `#include` is part of the `link` [preprocessor](https://rust-lang.github.io/mdBook/format/configuration/preprocessors.html#configuring-preprocessors) and it doesn't have access to the HTML configuration. It's possible to build a different preprocessor that does the "same", but adding your `src` at the preprocessor configuration (or change a lot the current one and add the prefix path. I guess that the easiest way to do it will apply it to all "`link` tags" (`{{ #playground }}`, `{{ #include }}`, and `{{ #rustdoc_include }}`)).

If I have some free time I can implement this preprocessor or try to extend/change the current `link` preprocessor.

**Comment by schungx** [comment]
The `#include` tag is what I use to automatically include a file containing URL links in abstract form, so that I can use simple link tags everywhere.

For example, I have `links.md`:

```md
[`Foo`](/path/to/foo.md)
[`Bar`](/path/to/bar.md)
[baz](/path/to/baz.md)
```

And then in each file:

```md
{{ #include ../links.md }}

Now I can use [`Foo`] and [`Bar`] and [baz] anywhere I like!
```

It works very well, except for having to always put in the right number of `../` to reach that particular file.

**Comment by watsom27** [comment]
Is there any update on this? 

The PR/Branch proposed seems like a decent fix, and would resolve the issues im having too, but this seems to have died in the water? 

Thanks

**Comment by watsom27** [comment]
Bump please @ehuss

Thanks

**Comment by calvinbrewer** [comment]
Any updates @ehuss? The proposed fix seems reasonable.

**Comment by daprilik** [comment]
Bump here. Ran into this just now, and needed to revert a PR that seemed to work fine in local development, but failed upon being deployed to github pages (where our book is hosted under `/guide/`).

**Comment by wilwade** [comment]
@daprilik This might be due to the changes in https://github.com/rust-lang/mdBook/pull/2414

I also had some issues and had to roll back to v0.4.40. Still trying to pin down exactly what the issue is, but before v0.4.41 the links in the sidebar are all relative. v0.4.41 made them non-relative, but (speculation) doesn't prepend the `base_url` / `site-url` as it should.

**Comment by commonsensesoftware** [comment]
Like @schungx, I have a file that has a bunch of links, primarily for shared definitions throughout the book; essentially a pseudo index. These can't be relative because they sit at the root and are referenced everywhere else.

@ehuss, I like @joaofreires's  idea, but I'm not convinced the flag is necessary nor is it a breaking change. If you do **not** specify `site-url`, then the default root is already `/`. Authors specify links like this:

- **Relative** - `chapter/page.md`
- **Absolute** - `/chapter/page.md`

An absolute link should **always** anchor from the root. When someone specifies `site-url` with a value other than `/`, then that should be root.

```toml
[output.html]
site-url = "/my-book/"
```

As an author, when I do this, I've indicated the root is now `/my-book` and I expect absolute URLs, such that `/chapter/page.md` becomes `/my-book/chapter/page.md`. Anything else feels like a [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) violation. This is further evident by the fact that everything _looks_ fine while authoring and proofing content locally, but as soon as you publish, many links are unexpectedly broken. Conversely, if you qualify the links as `/my-book/chapter/page.md`, they work when published, but all break locally.

I _could_ live with gating behind a feature (as proposed), but I don't think that's even necessary. In the event there's an edge case that I've missed, the other option would be to use `~`, which some web platforms support. This would mean that `~/chapter/page.md` would expand to `/my-book/chapter/page.md` or whatever the configured `site-url` is, if any. That would be backward compatible without toggling a feature.

**Comment by schungx** [comment]
Seeing that version 0.5 is coming out... I wonder if there is now a solution to this issue?

**Comment by JesusPerez** [comment]
@rustbot claim

Still reproducible on current `master`: with `output.html.site-url` set, only the 404 page uses it, so a book hosted in a subdirectory gets incorrect navigation, sidebar, asset and in-content links.

Implemented in #3139 — `site-url` now roots every generated link (navigation, the JS/no-JS sidebar, static assets, in-content `./` links and the print page) at the configured URL, with tests and docs.

It supersedes #1802 (reimplemented on the `crates/` layout, review addressed) and adds nothing to the public API (`site_url` is already a field on `HtmlConfig`).
